packages/datetime/src/datePicker.tsx

```diff
@@ -221,153 +186,146 @@ export class DatePicker extends AbstractPureComponent<IDatePickerProps, IDatePic
                 this.ignoreNextMonthChange = true;
             }
-        } else {
-            // rerender base component to get around bug where you can navigate past bounds by clicking days
-            this.forceUpdate();
+            this.updateValue(newValue, true);
         }
     };
 
     private computeValidDateInSpecifiedMonthYear(displayYear: number, displayMonth: number): Date {
         const { minDate, maxDate } = this.props;
+        // month is 0-based, date is 1-based. date 0 is last day of previous month.
         const maxDaysInMonth = new Date(displayYear, displayMonth + 1, 0).getDate();
-        let { selectedDay } = this.state;
-
-        if (selectedDay > maxDaysInMonth) {
-            selectedDay = maxDaysInMonth;
-        }
-
-        // matches the underlying react-day-picker timestamp behavior
-        let value = new Date(displayYear, displayMonth, selectedDay, 12);
+        const selectedDay = Math.min(this.state.selectedDay, 1, maxDaysInMonth);
 
+        // 12:00 matches the underlying react-day-picker timestamp behavior
+        const value = new Date(displayYear, displayMonth, selectedDay, 12);
+        // clamp between min and max dates
         if (value < minDate) {
-            value = minDate;
+            return minDate;
         } else if (value > maxDate) {
-            value = maxDate;
+            return maxDate;
         }
-
         return value;
     }
 
+    private handleClearClick = () => this.updateValue(null, true);
+
     private handleMonthChange = (newDate: Date) => {
         const displayMonth = newDate.getMonth();
         const displayYear = newDate.getFullYear();
-        let { value } = this.state;
-
-        if (value !== null) {
-            value = this.computeValidDateInSpecifiedMonthYear(displayYear, displayMonth);
-            if (this.ignoreNextMonthChange) {
-                this.ignoreNextMonthChange = false;
-            } else {
-                // if handleDayClick just got run, it means the user selected a date in a new month,
-                // so don't run onChange again
-                Utils.safeInvoke(this.props.onChange, value, false);
-            }
+        this.setState({ displayMonth, displayYear });
+
+        if (this.state.value !== null) {
+            const value = this.computeValidDateInSpecifiedMonthYear(displayYear, displayMonth);
+            Utils.safeInvoke(this.props.dayPickerProps.onMonthChange, value);
+            // if handleDayClick just got run (so this flag is set), then the
+            // user selected a date in a new month, so don't invoke onChange a
+            // second time
+            this.updateValue(value, false, this.ignoreNextMonthChange);
+            this.ignoreNextMonthChange = false;
         }
-
-        Utils.safeInvoke(this.props.dayPickerProps.onMonthChange, value);
-
-        this.setStateWithValueIfUncontrolled({ displayMonth, displayYear }, value);
+        // don't change value if it's empty
```

Comment:
nit: same. clearer structure IMO:

```tsx
if (this.state.value === null) {
    // don't change if it's empty
    return;
}

const value = this.computeValidDateInSpecifiedMonthYear(displayYear, displayMonth);
// ...
```

Reply:
i don't think that helps in this case where there's only one conditional, but it's great in the others above